File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    # 이번달 까지 두 사람 사이에 더 많은 선물을 준 사람이 다음달에 선물을 하나 받음
    # gifts = [A B, ... B C] : A 선물을 준 사람, B 받은 사람
    # 선물을 주고 받은 횟수 카운트
    # 선물을 주고 받은 횟수를 2차원의 배열로
    N = len(friends)
    cnt = [[0]*N for _ in range(N)]
    for i in range(len(gifts)):
        A, B = gifts[i].split()
        r, c = friends.index(A), friends.index(B)
        
        cnt[r][c] += 1
        cnt[c][r] -= 1
    
    result = [0]*N
    for i in range(N):
        for j in range(N):
            if i == j: continue
            
            if cnt[i][j] > 0: result[i] += 1
            elif not cnt[i][j] and sum(cnt[i]) > sum(cnt[j]): result[i] += 1
    
    # 선물을 가장 많이 받을 친구가 받을 선물의 수
    answer = max(result)
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

int solution(vector<string> friends, vector<string> gifts) {
    int answer = 0;
    int MAX = friends.size();
    map<string, int> m;
    int arr[MAX][MAX];
    memset(arr, 0, sizeof(arr));
    for (int i=0; i<MAX; i++)
        m[friends[i]] = i;
    
    int cnt[MAX];
    memset(cnt, 0, sizeof(cnt));
    string x, y;
    for (auto q : gifts){
        stringstream ss(q);
        ss >> x >> y;
        arr[m[x]][m[y]]++;
        cnt[m[x]]++;
        cnt[m[y]]--;
    }
    
    int result[MAX];
    memset(result, 0, sizeof(result));
    
    for (int i=0; i<MAX; i++){
        for (int j=0; j<MAX; j++){
            if (arr[i][j] > arr[j][i]) result[i]++;
            else if (arr[i][j] == arr[j][i] && cnt[i] > cnt[j]) result[i]++;
        }
    }
    
    for (auto r : result)
        answer = max(answer, r);
    return answer;
}